File: matlab/functions/vic_adaptive.m
function [P_vic, H_t, D_t, P_filt_new] = vic_adaptive(domega, domega_dt, P_filt_prev, params)
% VIC_ADAPTIVE  Adaptive Virtual Inertia Control (tanh-saturated H, D)
%   Sumber: julia/src/models/vic_adaptive.jl
%
% Inputs:
%   domega      : Δω = ω - ω0 [rad/s]
%   domega_dt   : dω/dt        [rad/s^2]
%   P_filt_prev : filter state [W]
%   params      : struct dari pv_bess_vic_params()
%
% Outputs:
%   P_vic       : virtual inertia power [W]
%   H_t         : effective inertia    [s]
%   D_t         : effective damping    [pu]
%   P_filt_new  : updated filter state [W]

    % Adaptive gains (tanh saturation @ disturbance scale)
    H_t = params.H_min + (params.H_max - params.H_min) * ...
          tanh(params.gamma_J * abs(domega_dt) / params.omega_dot_ref);
    D_t = params.D_min + (params.D_max - params.D_min) * ...
          tanh(params.delta_D * abs(domega) / params.omega_ref_vic);

    % Swing-equation form
    P_raw = -(2 * H_t / params.omega0 * domega_dt + ...
              D_t * domega / params.omega0) * params.P_rated;

    % First-order low-pass filter (T = T_filter_omega)
    alpha_filt = params.dt / (params.T_filter_omega + params.dt);
    P_filt_new = (1 - alpha_filt) * P_filt_prev + alpha_filt * P_raw;

    % Saturate at ±P_vic_max
    P_vic = max(min(P_filt_new, params.P_vic_max), -params.P_vic_max);
end

File: julia/src/models/vic_adaptive.jl
"""
Adaptive Virtual Inertia Control.

Outputs the active-power adjustment P_vic [W] in response to frequency
deviation Δω [rad/s] and rate of change dω/dt.

Inertia and damping gains are NOT constant — they adapt to disturbance
severity (key novelty over Tian et al. 2023 which used constant-output
adaptive logic only on damping):

    H(t) = H_min + (H_max - H_min) * tanh(γ_J * |dω/dt| / ω0)
    D(t) = D_min + (D_max - D_min) * tanh(δ_D * |Δω|   / ω0)

The synchronous-machine swing-equation analogue gives:

    P_vic = -[ 2H(t)/ω0 * dω/dt + D(t) * Δω/ω0 ] * P_rated

with output filtered (Tdi lead, Tri lag) and saturated to ±P_vic_max.
"""
@with_kw mutable struct AdaptiveVICState
    P_filt::Float64 = 0.0   # post-filter output
    Δω_lp::Float64  = 0.0   # low-pass of Δω for derivative estimation
end

function vic_output(Δω::Float64, dωdt::Float64, p::SystemParams,
                    s::AdaptiveVICState; dt::Float64=p.Ts_control)
    # adaptive gains — tanh dinormalisasi ke RoCoF/Δf tipikal supaya saturasi
    # tercapai pada event sedang (bukan ke ω0 yang membuat arg<0.05 dan H_t≈H_min)
    H_t = p.H_min + (p.H_max - p.H_min) * tanh(p.gamma_J * abs(dωdt) / p.omega_dot_ref)
    D_t = p.D_min + (p.D_max - p.D_min) * tanh(p.delta_D * abs(Δω)   / p.omega_ref_vic)

    # raw VIC contribution (negative sign: counteract deviation)
    P_raw = -(2*H_t/p.omega0 * dωdt + D_t * Δω/p.omega0) * p.P_rated

    # first-order shaping filter (T_lead RoCoF + T_lag droop) approximated
    # as a single low-pass (T_filter_omega) for the explicit Euler step
    α_filt = dt / (p.T_filter_omega + dt)
    s.P_filt = (1 - α_filt) * s.P_filt + α_filt * P_raw

    # saturation
    P_out = clamp(s.P_filt, -p.P_vic_max, p.P_vic_max)
    return P_out, (H=H_t, D=D_t)
end
